ruby/object_oriented_programming_basics/linting_and_rubocop.md

```diff
@@ -0,0 +1,307 @@
+
+<!--
+TODO: find out whether https://docs.rubocop.org/rubocop/usage/autocorrect.html talks about human semantics or the machine semantics.
+The second example seems to suggest the latter - autocorrecting breaks the code.
+-->
+The latter means that while your code and the proposed code arrive at the same output, they might change how (but not what) the output is achieved or produce a side-effect that might even break your code.
+
+### You are the code dictator
+
+Due to Ruby's ecosystem, RuboCop was built with extensive configurability in mind - both in terms of not using some parts of and in terms of adding onto it. Every single Cop can be disabled, sometimes Cops offer alternative rules like preferring single- or double-quotes for Strings, you can disable Cops on a per-file basis and much more.
+
+Since RuboCop is extensible, there exist other departments that you can use - like Performance or RSpec. You could even write your own Cop! The process of adding an extension is easy: you install the Gem locally and modify `.rubocop.yml`.
+
+Usually things that are not required for app to run are given the `require: false` flag, like:
+
+```bash
+`gem 'rubocop-performance', require: false`
+```
+
+This way the Gem would be installed normally, but for your `bundle exec` ran code to make use of it, it would need to be explicitly `require`d wherever you'd need it.
+
+`.rubocop.yml` is the configuration file for RuboCop and it lives in the root directory of your project. There you'll change the defaults of RuboCop to your (or most likely, your team's) liking. To create such config file, you can just use `bundle exec rubocop --init` - it won't have anything in it besides a comment describing what it is for but if we were to add the Performance extension, we'd need to throw `require: rubocop-performance` in there so RuboCop knows to run it.
+
+RuboCop is still under development, so changes and additions happen. New Cops join the precint and they're not enabled by default - if you'd like them to be enabled by default instead of going through all of them and deciding on your own, you can use:
+
+```bash
+AllCops:
+  NewCops: enable
+```
+
+in your `.rubocop.yml` to enable all the new Cops.
+
+### Power corrupts
+
+Perhaps you're not interested in tailoring RuboCop to your liking, especially since you're just starting out and have absolutely no idea what's good and what's bad. That'd be the correct approach - don't worry about it right now and just go with the RuboCop defaults.
+
+There exists another set of rules which is executed by a wrapper over RuboCop called Standard Ruby. Standard aims at providing one opinionated set of rules, smaller in scope than RuboCop, in hopes that it will provide a standard for all Rubyists to follow.
```

Comment:
I'm not sure either. My rationale for the inclusion was mostly the fact that I could foresee people going into some deep rabbit hole of RuboCop configuration or grabbing config off some company just for the sake of having a config that is not default. This might have some bad consequences like in my mind like the lack of Metrics cops mentioned in the draft.

Buuut just mentioning Standard might do just that, even though I've made a (hopefully) strong case for sticking with RuboCop. I'm OK with dropping mention of Standard completely, so I wouldn't even put it in additional resources.

Reply:
I think the config rabbit hole can be avoided by doing two things:
1. Mentioning that people should stick to rubocop's defaults when starting out. You kind of have this going on already with the paragraph starting on lines 204, but I think it's good advice for people who are starting out
2. I think it'd also be potentially good to show how to configure the two most common things people have problems with on rubocop's default settings: the magic comment for frozen string literals and using double instead of single quotes for strings.

What do you think?

Reply:
So hammer the point mentioned in 1. harder + show how to throw magic comment/quote string rules into `rubocop.yml`?

Reply:
Yep!

I think this would help people avoid too much config rabbit holing. We can potentially revisit though if config bikeshedding seems to be a problem in discord conversations.